[HOPSWORKS-3259] Normalize API for reading changes based on commit time

tdoehmen opened the pull request:
- added `exclude_until` parameter to `fg.as_of` 
- marked `fg.read_changes` as deprecated
- same changes for Java client
- fixed format bug in `get_timestamp_from_date_string`

moritzmeister commented on python/hsfs/constructor/query.py:
```diff
@@ -318,12 +330,12 @@ def left_feature_group_start_time(self):
     @property
     def left_feature_group_end_time(self):
-        return self._left_feature_group_start_time
+        return self._left_feature_group_end_time
```
nice catch :)

moritzmeister reviewed (commented):
1. Why only deprecating in StreamFeatureGroup but not in CachedFeatureGroup?
2. We need documentation for the pull_changes, it's not clear to me what the difference is now compared to as_of with the exclude_until.

To be honest I am kind of confused @SirOibaf @tdoehmen. 

Initially, `as_of` was meant to map to the snapshot query in Hudi, you present the table to a user as of a single certain commit time, that's why the method has one wallclock_time.
`read_changes` was meant to map to the incremental query in Hudi, reading the changes between two commits.

@SirOibaf we were always using the incremental query though for both, I don't remember the original reason for it. And then you switched it the other day to snapshot query when no timestamp is provided to avoid the "active/inactive" commit timeline issue, right?

The original issue by Martin was about the fact that it's inconsistent when we are returning Dataframes vs. Queries, but the concept of reading changes vs time travel seemed to be clear. But now I cannot wrap my head around why we would deprecate read_changes, is it cause `pull_changes` will provide that functionality?
For `as_of` why would I want to specify a `exclude_until` if I want a snapshot?

TDLR; I am very confused. Maybe you guys have discussed it more when I was away.

moritzmeister commented:
Okay after a quick discussion with Till, it makes more sense. It's been decided like this, so it's okay. But I think mixing concepts in the same `as_of` method will lead to confusion.

tdoehmen commented:
There is another case we discussed yesterday:

- as_of with only start time: Incremental query that considers all the data committed from the start time to the last commit in the Hudi active timeline. 

I will need to make another commit to support this. 

And a remark on this:
- as_of with start and end time: Incremental query that considers all the data from the **last commit before start_time** to the last commit before the end_time. Both start and end need to be in the active timeline.

According to the hudi documentation, hoodie.datasource.read.begin.instanttime gets the **first commit after start_time** ("New data written with an instant_time > BEGIN_INSTANTTIME are fetched out. For e.g: ‘20170901080000’ will get all new data written after Sep 1, 2017 08:00AM.") 
I would argue that neither for the start_time nor for the end_time we have to manually lookup the closest commit, as Hudi will do the lookup implicitly when setting hoodie.datasource.read.begin.instanttime or hoodie.datasource.read.end.instanttime. 
Or is there a reason that we purpusefully want start_time to be **last commit before start_time** ?

One more thing we discussed yesterday was whether as_of should execute a snapshot query and end_time, when not start_time is set. If I understand correctly, in this case the incremental query and the snapshot query would yield the same result.

tdoehmen commented:
Made a few updates:

- made `as_of ` work with `wallclocktime/end_time=None`
- added `as.of.instant` readOps when only `end_time` is set (as suggested by Moritz)
- added deprecations warning to `pull_changes` 

The function name is still in debate. @moritzmeister @SirOibaf What shall we settle on?

The options A) as implemented in this PR (all behaviour in one function):
- `as_of(wallclock_time, exclude_until=None)`
    - when only wallclock_time is set -> calls as.of.instance query
    - exclude_until is set -> call incremental query

Option B) as suggested by Moritz (seperate as.of.instant and incremental queries):
- `as_of(wallclock_time)`
    - calls as.of.instant query 
- `pull_changes/from_to(start_timestamp=None, end_timestamp=None)`
    - calls incremental query

My opinion is that one function is sufficient and should be okay to understand if we document it properly (I could take over the docstrings from you suggestion branch @moritzmeister). We could rename the parameter of the `as_of` function to something more meaningful, such as `as_of(latest_timestamp, earliest_timestamp)` (at the price of breaking backwards compatability to users using names function arguments?)

moritzmeister commented:
> If I understand correctly, in this case the incremental query and the snapshot query would yield the same result.
With the slight difference that the incremental query takes only into account the active timeline.

I am fine with putting it into one function (as_of). Seems like you two outvoted me, so that's ok.

Can you add the comment in the docs about that it matters if you have joins at which point you add the .as_of call?

> 
> Note that the interval will be applied to all joins in the query.
> If you want to query different intervals for different feature groups in
> the query, you have to apply them in a nested fashion:
> 
> ```python
> fg_a.select_all().from_to(..., ...)
>     .join(fg_b.select_all().from_to(..., ...))
> ```
> 
> If instead you apply another `from_to` selection after the join, all
> joined feature groups will be queried with this interval:
> 
> ```python
> fg_a.select_all().from_to(..., ...)  # from_to is not applied
>     .join(fg_b.select_all().from_to(..., ...))  # from_to is not applied
>     .from_to(..., ...)
> ```

tdoehmen commented:
@moritzmeister Great, I added that also to the docs. 

@SirOibaf @moritzmeister Shall we still rename the params to `as_of(latest_timestamp, earliest_timestamp)` or leave it the way it is now?

moritzmeister commented:
@tdoehmen I would leave it the way it is so be don't have a breaking change between minor versions. We can do it for the 4.0 release.

tdoehmen commented:
Okay, I guess then we're ready to go.

moritzmeister reviewed (commented):
one more comment about the documentation

tdoehmen commented on python/hsfs/feature_group.py:
```diff
@@ -1209,21 +1214,80 @@ def commit_delete_record(
+            This method only works for feature groups with time time_travel_format=`HUDI`.
+
+        !!! warning
+            The wallclock_time needs to be a time included into the Hudi active timeline.
```
Let me add that
